### PR TITLE
Simplify Expr.arithmetic

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
- Fix https://github.com/ethereum/vyper/issues/1658
- Always precompute both sides into an LLL reference (pre_alloc for call
    no longer needed), so generally more gas efficient if the arguments
    are more than ~2 ops
- Remove a jumpi for mul uint256
- Simplify type calculation / pos calculation
- Use clamps instead of seq-assert where it's clearer
- Use similar styles for similar output (e.g. mul uint256 looks like mul
    int128)

### How I did it
Use references to break the exponential behavior in #1658 

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Comment by charles-cooper** [comment]
and of course that breaks constant folding.

**Comment by jacqueswww** [comment]
@charles-cooper it's not breaking constant folding, you just have to fix the test to extract the constant value differently now that the LLL output has changed ;)

File: vyper/parser/expr.py [hunk 1]
```diff
@@ -521,117 +505,95 @@ def arithmetic(self):
...
+                                     ['iszero', 'l']]],
+                             'ans']]
+
             elif ltyp == rtyp == 'int128':
-                o = LLLnode.from_list(
-                    ['mul', left, right],
-                    typ=BaseType('int128', new_unit),
-                    pos=getpos(self.expr),
-                )
+                # TODO should this be 'smul' (note edge cases in YP for smul)
```
Comment (fubuloubu) [inline vyper/parser/expr.py]: Let's fix this TODO

File: vyper/parser/expr.py [hunk 2]
```diff
@@ -521,117 +505,95 @@ def arithmetic(self):
...
-                    'with', 'l', left, [
-                        'with', 'r', ['clamp_nonzero', right], [
-                            'sdiv',
-                            ['mul', 'l', DECIMAL_DIVISOR],
-                            'r',
-                        ],
-                    ]
-                ], typ=BaseType('decimal', new_unit), pos=getpos(self.expr))
+                arith = ['sdiv',
+                         # TODO check overflow cases
```
Comment (fubuloubu) [inline vyper/parser/expr.py]: Let's fix this TODO

Comment (charles-cooper) [reply vyper/parser/expr.py]: i think it's covered by the clamp later actually

File: vyper/parser/expr.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/parser/expr.py]: ditto about the later clamp

Comment (charles-cooper) [reply vyper/parser/expr.py]: @fubuloubu I don't want to change semantics in this PR, TODO is a good flag for somebody to look at it later.

**Pull request merged**